llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -6505,6 +6505,33 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
 
     break;
   }
+  case Intrinsic::scmp:
+  case Intrinsic::ucmp: {
+    // Fold cmp x, x -> 0
+    if (Op0 == Op1)
+      return Constant::getNullValue(ReturnType);
```

Comment:
This is subsumed by the following case.

Reply:
Seems like a reasonably enough fast path though, no?

Reply:
Yeah, I kept it there because I assumed that `isICmpTrue` would be much slower than just comparing two pointers, even if `isICmpTrue` has this fast path as well, just burried a few calls deep.

Reply:
The rule of thumb is that you only need to optimize the case where things do *not* fold, because not folding is the common case. Unless a substantial fraction of inputs contain cmp(x, x), having a special case for it will make no difference.

If we wanted to optimize this, then we could drop the ICMP_EQ query instead, and argue that cases where the values are non-trivially equal should be handled by GVN instead. But I think it's fine to keep the general implementation for now.

Reply:
I get your point, however I think that another case where this may be helpful is that it works for both scalar and vector types, unlike the code involving `isICmpTrue`. However, I'm happy to delete it if it is deemed to be unnecessary anyway.

Reply:
isICmpTrue() also works on vector types, or at least it's supposed to...

Reply:
Yep, you're right, I've removed the check.